controllers/metal3.io/preprovisioningimage_controller.go

```diff
@@ -110,6 +110,13 @@ func (r *PreprovisioningImageReconciler) Reconcile(ctx context.Context, req ctrl
 		log.Info("requeuing to check for secret", "after", delay)
 		result.RequeueAfter = delay
 	}
+
+	notReady := imageprovider.ImageNotReady{}
+	if errors.As(err, &notReady) {
+		log.Info("image is not ready yet, requeuing", "after", notReady.RequeueAfter)
+		result.RequeueAfter = notReady.RequeueAfter
+	}
```

Comment:
This is not updating the status to say that the image is not ready, just leaving the CR in whatever state it was. I think you also/instead need to check for this right after the check for ImageBuildInvalid.

Reply:
I'm not sure I'm doing the right thing, but I've tried. Please check.

Reply:
I'm definitely not changing the whole `BuildImage` signature for one error :-/ This value is specific to the error and should be there, but if you insist, I'll remove it and just have a constant retry.

As to your first comment, we don't have an access to `Result` in `update`, thus I cannot move the check there.

Reply:
IMHO a custom error type should contain only the strictly necessary additional fields to report the error, help the troubleshooting and the debugging activity. 
The current retry approach seems ok to me (as the request will be automatically re-enqueued with a rate limiting), but in case I'd consider ok to review the `ImageProvider` api if required.

/lgtm

Reply:
This looks fine, but the only reason for having the code here was because it was setting the retry delay from data in the error. I wasn't opposed to doing that, but if we are not going to then this code definitely belongs at the point where the error is first returned, i.e. https://github.com/metal3-io/baremetal-operator/blob/05d12b6768a9989a9a4e61dad6cd1f9e84a6e078/controllers/metal3.io/preprovisioningimage_controller.go#L186